Spigot-Server-Patches/0675-Fix-hopper-lag-by-making-entities-look-for-hoppers.patch

```diff
@@ -0,0 +1,309 @@
+         }
+     }
+ 
++    private final boolean isInventoryFull() { return j(); } // Paper - OBFHELPER
+     private boolean j() {
+         Iterator iterator = this.items.iterator();
+ 
+@@ -320,10 +327,20 @@ public class TileEntityHopper extends TileEntityLootable implements IHopper, ITi
+             ((EntityMinecartHopper) hopper).setCooldown(hopper.getWorld().spigotConfig.hopperTransfer / 2);
+         }
+     }
++
++    public AxisAlignedBB getHopperLookupBoundingBox() {
++        // Change this if b(IHopper) ever changes
++        return getHopperLookupBoundingBox(this.getX(), this.getY() + 1.0D, this.getZ());
++    }
++
++    private static AxisAlignedBB getHopperLookupBoundingBox(double d0, double d1, double d2) {
++        // Change this if the above ever changes
++        return new AxisAlignedBB(d0 - 0.5D, d1 - 0.5D, d2 - 0.5D, d0 + 0.5D, d1 + 0.5D, d2 + 0.5D);
++    }
+     // Paper end
+ 
+     private boolean k() {
+-        IInventory iinventory = this.l();
++        IInventory iinventory = getInventory(getWorld(), getPosition().shift((this.getBlock().get(BlockHopper.FACING)))); // Paper
+ 
+         if (iinventory == null) {
+             return false;
+@@ -431,8 +448,21 @@ public class TileEntityHopper extends TileEntityLootable implements IHopper, ITi
+     }
+ 
+     public static boolean a(IHopper ihopper) {
+-        IInventory iinventory = b(ihopper);
++        // Paper start
++        IInventory iinventory;
++        if (ihopper.getWorld().paperConfig.isHopperPushBased
++            && ihopper instanceof TileEntityHopper) {
++            BlockPosition pos = ((TileEntityHopper) ihopper).getPosition()
++                .up(); // Only pull from a above, because everything else comes to us
++            iinventory = getInventory(ihopper.getWorld(), pos);
++        } else {
++            iinventory = getSourceInventory(ihopper); // Use old behavior for BB entity searching
++        }
++        return acceptItem(ihopper, iinventory);
++    }
+ 
++    public static boolean acceptItem(IHopper ihopper, IInventory iinventory) {
++    // Paper end
+         if (iinventory != null) {
+             EnumDirection enumdirection = EnumDirection.DOWN;
+ 
+@@ -447,7 +477,7 @@ public class TileEntityHopper extends TileEntityLootable implements IHopper, ITi
+                 }
+             });
+             // Paper end
+-        } else {
++        } else if (!ihopper.getWorld().paperConfig.isHopperPushBased || !(ihopper instanceof TileEntityHopper)) { // Paper - only search for entities in 'pull mode'
+             Iterator iterator = c(ihopper).iterator();
+ 
+             EntityItem entityitem;
+@@ -458,10 +488,11 @@ public class TileEntityHopper extends TileEntityLootable implements IHopper, ITi
+                 }
+ 
+                 entityitem = (EntityItem) iterator.next();
+-            } while (!a((IInventory) ihopper, entityitem));
++            } while (!canPickupItem((IInventory) ihopper, entityitem));
```

Comment:
A Paper comment there as well? Or instead use the non obfhelper method there and reduce the patch size?